### PR TITLE
form conditions fields

### LINKED ISSUE #8595: dependent form fields
# Description
For the marketer, it is very useful to be able to display a form field depending on the value of a previous field within the same form.

## Example
You have a form to display car brands and car models. The model list should depend on the brand selected by the user.
* If `Renault` in brands field, display in the model list: `Clio, Megane, Trafic, etc.`
* If `Volkswagen` in brands field, display in the model list: `Polo, Golf, Transporter, etc.`

## Demo
Capture d’écran 2020-03-26 à 11 16 16

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/90653563-dependent-form-fields?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by escopecz**
Documentation PR: https://github.com/mautic/mautic-documentation/pull/208

~⚠️ This PR is blocked by #9761 which needs to be merged before this PR.~

Clone of https://github.com/mautic/mautic/pull/8660 for M3

Closes #8595 

A new feature was requested for combination of Country and conditional State field. The goal is to show only the states for the selected country.

### Steps to test
- Follow https://github.com/mautic/mautic/pull/8660

### To test the Company-state field feature:
1. Create a new form
2. Create a country field
3. Create a new conditional select field for the country field. Make it visible for any country value and map it to the State contact field. The choices should be used from the contact field as well. Do not create custom ones.
4. Save the form.
5. Go to the form preview.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/8794"></a>

### CODE REVIEW

**Comment by npracht** [comment]
Link to doc : https://docs.google.com/document/d/1JfLmimbuYPXRwxBrUcusM1-ITQLP-TyCMU1QGvG27Ss/edit?ts=5ec7d494#heading=h.vc029lot0qwv

**Comment by kuzmany** [comment]
I found issue for M2 version If parent field is hidden type (Show If value exist is disabled)

https://github.com/mautic/mautic/pull/8660/commits/e40aaf33a217dbe0b97275720d9d88bb4a45c6ad

**Comment by escopecz** [comment]
@kuzmany thinking about this, is there even any use case why we should make hidden conditional fields? The hidden field will be in the HTML for the whole time anyway. No matter whether visible or hidden.

**Comment by kuzmany** [comment]
@escopecz 
we have two states of hidden

1. If showForContact = false https://github.com/mautic/mautic/blob/staging/app/bundles/FormBundle/Views/Builder/form.html.php#L67
Then input is not render to form
2. If there is some combinations getShowWhenValueExists, getLeadField, getIsAutoFill ...
https://github.com/mautic/mautic/blob/staging/app/bundles/FormBundle/Views/Builder/form.html.php#L77, then is input covert to hidden type

For the second one , we don't wanna generate children field.

File: app/bundles/FormBundle/Views/Field/freehtml.html.php
Comment (kuzmany) [inline app/bundles/FormBundle/Views/Field/freehtml.html.php]: This was removed from my commits or from yours? 
This should prevent run script code from builder
We could revert it

image

**Review by kuzmany — CHANGES_REQUESTED** [review]

File: app/bundles/FormBundle/Views/Field/freehtml.html.php
Comment (escopecz) [reply app/bundles/FormBundle/Views/Field/freehtml.html.php]: Good catch. I will push a fix shortly together with other bug we found.

Comment (escopecz) [reply app/bundles/FormBundle/Views/Field/freehtml.html.php]: To answer your question, this is the commit that removed it: https://github.com/mautic/mautic/commit/1293c0f470b769d4c6a9b8f1a9cc364f96688de6#diff-0ffe1ecc742a8abb035d9e423721e222R15

Comment (kuzmany) [reply app/bundles/FormBundle/Views/Field/freehtml.html.php]: Yes, I found it. But didn't have time check it why I remove it. Maybe accidentaly

Comment (escopecz) [reply app/bundles/FormBundle/Views/Field/freehtml.html.php]: Fixed

**Comment by escopecz** [comment]
This PR is blocked by our [internal PR](https://github.com/mautic-inc/mautic-cloud/pull/838) that we cannot push to the community because it is blocked by https://github.com/mautic/mautic/pull/8828

**Comment by kuzmany** [comment]
We've added support for 

- checkbox groups https://github.com/mautic/mautic/pull/8660/commits/028ce4f08e477f374e648b291cf33172c4524066
- radio groups https://github.com/mautic/mautic/pull/8660/commits/79fea12238bfc3a0a7a74915dc2a435f0846f716

Still consider add fix
- https://github.com/mautic/mautic/pull/8660/commits/e40aaf33a217dbe0b97275720d9d88bb4a45c6ad

reported here https://github.com/mautic/mautic/pull/8794#issuecomment-635214289

**Comment by escopecz** [comment]
@kuzmany Thanks! I cherry-picked those commit and added to this PR.

**Comment by RCheesley** [comment]
Close & open for CI

**Comment by RCheesley** [comment]
@escopecz could you take a look at Travis errors please?

**Comment by escopecz** [comment]
@RCheesley I added a note why it's failing into a comment. But the history of this PR is too long and it was hidden. I added it as a warning to the PR description. This PR is dependent on 2 other open PRs.

**Comment by RCheesley** [comment]
Here's hoping we can get them tested and documented so that they can be merged in 3.1! Most are features so that's the earliest chance we have to merge them.

It would be great if the Acquia team could please take a look at their PR's at some point as many are missing documentation, have conflicts, need to be rebased etc.

**Comment by RCheesley** [comment]
Bumping to 3.2 due to PR that blocks this not having responses to feedback. Here's hoping we can get these PR's merged in November!

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by npracht** [comment]
@kuzmany can you review it ? It would be good to merge it in 3.3.0 :)

**Comment by RCheesley** [comment]
@mautic/acquia-po @escopecz @alanhartless the internal PR should now be unblocked - would be helpful if it could be pushed over to the community so that we can get this feature released!

**Comment by kuzmany** [comment]
Related fixes to this PR https://github.com/escopecz/mautic/pull/17

File: app/bundles/CoreBundle/Assets/js/4.builder.js
Comment (kuzmany) [inline app/bundles/CoreBundle/Assets/js/4.builder.js]: I see here some changes what not related to this PR as like this

**Review by kuzmany — CHANGES_REQUESTED** [review]
media/css/app.css 
media/js/app.js
mautic-form.js 

should not be part of this PR

**Comment by RCheesley** [comment]
@fedys could you take a look at the feedback from @kuzmany above? Thanks!

File: app/bundles/CoreBundle/Assets/js/4.builder.js
Comment (fedys) [reply app/bundles/CoreBundle/Assets/js/4.builder.js]: @kuzmany this change was made by you. Shall I revert it?

Comment (kuzmany) [reply app/bundles/CoreBundle/Assets/js/4.builder.js]: Yes

File: app/bundles/LeadBundle/EventListener/ReportSubscriber.php
Comment (kuzmany) [inline app/bundles/LeadBundle/EventListener/ReportSubscriber.php]: This should not part of this PR

File: plugins/MauticFocusBundle/EventListener/CampaignSubscriber.php
Comment (kuzmany) [inline plugins/MauticFocusBundle/EventListener/CampaignSubscriber.php]: Shouln't part of this PR

File: .github/workflows/mautic-asset-upload/upload.php
Comment (kuzmany) [inline .github/workflows/mautic-asset-upload/upload.php]: Shouln't part of this PR

File: app/bundles/ConfigBundle/Controller/ConfigController.php
Comment (kuzmany) [inline app/bundles/ConfigBundle/Controller/ConfigController.php]: Shouln't part of this PR

**Review by kuzmany — CHANGES_REQUESTED** [review]
I found more files not related to this PR, I commented it.

**Review by kuzmany — DISMISSED** [review]
We use this on production long time. It works as describe 👍

**Comment by RCheesley** [comment]
@fedys @escopecz there are still files in this PR that should not be, according to the comments from @kuzmany.

Could you please take a look and address those issues and the failing tests? Thanks!

**Comment by RCheesley** [comment]
I have written the documentation PR based on the PR from @kuzmany and the notes on this PR: https://github.com/mautic/mautic-documentation/pull/208

**Comment by RCheesley** [comment]
@escopecz @ts-navghane this one is blocking a few other PRs - could we address the conflict so we can get this merged in 4.1? Thank you in advance!

**Comment by RCheesley** [comment]
Blocking PR is now merged so hopefully the tests will start to pass (we may need to pull the 4.0 branch up to the 4.x branch first) and we can start to test this PR, which is blocking a *lot* of other PRs at Acquia from coming over to the community.

**Comment by RCheesley** [comment]
@escopecz the PR this was dependent on is now merged, and should be within this branch. Hopefully we can get cracking with testing this PR once the tests are passing.

File: .github/workflows/mautic-asset-upload/upload.php
Comment (kuzmany) [inline .github/workflows/mautic-asset-upload/upload.php]: Should not be this part of this PR  I think

Comment (escopecz) [reply .github/workflows/mautic-asset-upload/upload.php]: @fedys I'm also confused what is it doing here. I can see it's coming from https://github.com/mautic/mautic/commit/d983a889681096a25f2674f6ed63689ddf337add. Do you remember what is the story here?

Comment (fedys) [reply .github/workflows/mautic-asset-upload/upload.php]: @escopecz I think I was just fixing some PR that included unrelated changes.

Comment (kuzmany) [reply .github/workflows/mautic-asset-upload/upload.php]: There are more changes not related to this PR for example:

https://github.com/mautic/mautic/pull/8794/files#diff-e192a95a44e25a558822230a82b49a099f01a1f2a6ffecf274b06b1579e7a94cR36
https://github.com/mautic/mautic/pull/8794/files#diff-fb63f85c5c793dbfedafce713ba68125484bd35d7cab18c3e0493a359a26cb8bL222

**Comment by escopecz** [comment]
@kuzmany I reverted changes that shouldn't be in this PR from https://github.com/mautic/mautic/pull/8794/commits/d983a889681096a25f2674f6ed63689ddf337add by https://github.com/mautic/mautic/pull/8794/commits/00a84d79a83b85b3a21dcbfef19bd72d81436ecf

**Comment by kuzmany** [comment]
@escopecz thanks, I will check it. Don't forget we have another fix in your repo https://github.com/escopecz/mautic/pull/18

**Comment by escopecz** [comment]
I'm stuck on resolving the failing tests here. Here are my observations:
1. The tests are failing only in GitHub Actions.
2. The tests  pass fine on my local if I run only the specific test as well as if I run the whole test suite.
3. It was passing yesterday on https://github.com/mautic/mautic/pull/8794/commits/05171eea6b264af911253b31f15b1e56ef7e0f57 but it started to fail on commit that removed some code that was not related to this PR (https://github.com/mautic/mautic/pull/8794/commits/00a84d79a83b85b3a21dcbfef19bd72d81436ecf). I cannot find any link between the failing test and removed code.
4. I tried to run the failing tests in a separate process as it sometimes helps in these cases, but it did not.

@fedys @dennisameling would you be able to check if you'll get more ideas what else to try?

**Comment by escopecz** [comment]
The failing tests need fixes from https://github.com/mautic/mautic/pull/10470, so waiting for that to be merged first.

File: app/bundles/CoreBundle/Assets/js/4.builder.js
Comment (kuzmany) [inline app/bundles/CoreBundle/Assets/js/4.builder.js]: This return error with image insert by legacy builder

image

**Review by kuzmany — DISMISSED** [review]
I did basic test and works. 👍 
Code looks 👍 
Covered by unit tests 👍 
Let's go 👍 

I see just issue in app/bundles/CoreBundle/Assets/js/4.builder.js

File: app/bundles/CoreBundle/Assets/js/4.builder.js
Comment (escopecz) [reply app/bundles/CoreBundle/Assets/js/4.builder.js]: Good catch @kuzmany! Fixed in https://github.com/mautic/mautic/pull/8794/commits/0206837cd4bac7f1ded577da369d24ac71978b09

**Comment by dennisameling** [comment]
> The failing tests need fixes from #10470, so waiting for that to be merged first.

@escopecz FYI #10470 has just been merged, so that should unblock this PR 👍🏼

**Comment by escopecz** [comment]
Thanks @dennisameling! I cherry-picked the commits I needed earlier today. So this PR is not failing anymore.

**Review by kuzmany — DISMISSED** [review]
Very very cool PR 
Tested, unit tests included 
Then, let's go 👍

**Comment by RCheesley** [comment]
@escopecz looks like we have a conflict to address here now.

**Review by bell87 — DISMISSED** [review]
Tested, works well.

**Comment by RCheesley** [comment]
GTG as soon as the conflict is resolved! 🎉

**Comment by escopecz** [comment]
Conflict resolved

**Comment by RCheesley** [comment]
Thanks for all the work on this PR and persistence over the last year folks!

**Comment by RCheesley** [comment]
@all-contributors please add @bell87 for usertesting

**Pull request merged**
